File: 2.py
def calculate_steps(n, flowers, capacity):
    steps = 0
    current_pos = -1  # Начинаем у реки
    current_water = capacity
    
    for i in range(n):
        # Перемещаемся к текущему цветку
        steps += abs(i - current_pos)
        current_pos = i
        
        remaining = flowers[i]
        while remaining > 0:
            # Поливаем часть цветка
            water = min(current_water, remaining)
            remaining -= water
            current_water -= water
            
            # Проверяем необходимость возврата к реке
            need_refill = False
            if remaining > 0:
                need_refill = True
            elif i < n - 1 and current_water < flowers[i + 1]:
                need_refill = True
            
            if need_refill:
                # Возвращаемся к реке
                steps += abs(current_pos - (-1))
                current_pos = -1
                current_water = capacity
    
    return steps

# Ввод данных с клавиатуры
n = int(input("Введите количество цветов: "))
flowers_input = input("Введите потребности цветов через пробел или запятую: ").replace(',', ' ')
flowers = list(map(int, flowers_input.split()))
flowers = flowers[:n]  # Обрезаем до указанного количества цветов
capacity = int(input("Введите вместимость лейки: "))

# Расчёт и вывод результата
print("Количество шагов:", calculate_steps(n, flowers, capacity))

File: 2.rb
def calculate_steps(n, flowers, capacity)
  steps = 0
  current_pos = -1  # Начинаем у реки
  current_water = capacity

  n.times do |i|
    # Перемещаемся к текущему цветку
    steps += (i - current_pos).abs
    current_pos = i

    remaining = flowers[i]
    while remaining > 0
      # Поливаем часть цветка
      water = [current_water, remaining].min
      remaining -= water
      current_water -= water

      # Проверяем необходимость возврата к реке
      need_refill = remaining > 0 || (i < n - 1 && current_water < flowers[i + 1])

      if need_refill
        # Возвращаемся к реке
        steps += (current_pos - (-1)).abs
        current_pos = -1
        current_water = capacity
      end
    end
  end

  steps
end

# Ввод данных
print "Введите количество цветов: "
n = gets.chomp.to_i

print "Введите потребности цветов через пробел или запятую: "
flowers_input = gets.chomp.gsub(/,/, ' ').split
flowers = flowers_input.map(&:to_i)[0...n]

print "Введите вместимость лейки: "
capacity = gets.chomp.to_i

# Расчёт и вывод результата
puts "Количество шагов: #{calculate_steps(n, flowers, capacity)}"
